flowmachine/flowmachine/features/subscriber/active_subscribers.py

```diff
@@ -0,0 +1,181 @@
+
+
+    >>>     active_subscribers = ActiveSubscribers(
+                start_date=date(year=2016, month=1, day=1),
+                end_date=date(year=2016, month=1, day=2),
+                minor_period_threshold=3,
+                major_period_threshold=1,
+                tables=["events.calls"],
+            )
+
+    Returns subscribers who were active in at least five hours, at least three days out of the four
+    between 2016-01-01 and 2016-01-04
+
+
+    >>>     active_subscribers = ActiveSubscribers(
+                start_date=date(year=2016, month=1, day=1),
+                end_date=date(year=2016, month=1, day=4),
+                minor_period_threshold=5,
+                major_period_threshold=3,
+                tables=["events.calls"],
+            )
+
+    Returns subscribers that were active in at least two ten minute intervals within half an hour,
+    at least three times across the two hours between 20:00:00 and 22:00:00 on 2016-01-01
+
+
+    >>> active_subscribers = ActiveSubscribers(
+                start_date=datetime(year=2016, month=1, day=1, hour=20),
+                end_date=datetime(year=2016, month=1, day=1, hour=22),
+                minor_period_threshold=2,
+                major_period_threshold=3,
+                tables=["events.calls"],
+                minor_period_count=3,
+                minor_period_length=10,
+                period_unit="minutes",
+            )
+
+
+    """
+
+    period_to_rrule_mapping = {
+        "days": rr.DAILY,
+        "hours": rr.HOURLY,
+        "minutes": rr.MINUTELY,
+    }
+
+    def __init__(
+        self,
+        start_date: Union[date, datetime, str],
+        end_date: Union[date, datetime, str],
+        minor_period_threshold: int,
+        major_period_threshold: int,
+        subscriber_identifier: Optional[str] = "msisdn",
+        tables: Optional[List[str]] = None,
+        subscriber_subset: Optional[SubscriberSubsetType] = None,
+        minor_period_count: int = 24,
+        minor_period_length: int = 1,
+        period_unit: Literal["days", "hours", "minutes"] = "hours",
+    ):
+        self.start_date = start_date
+        self.end_date = end_date
+        self.minor_period_threshold = minor_period_threshold
+        self.sub_id_column = subscriber_identifier
+        self.events_tables = tables
+        self.major_period_threshold = major_period_threshold
+
+        self.events_table_query = EventsTablesUnion(
+            self.start_date,
+            self.end_date,
+            tables=tables,
+            subscriber_identifier=subscriber_identifier,
+            columns=[subscriber_identifier, "datetime"],
+            subscriber_subset=subscriber_subset,
+        )
+
+        date_generator = rr.rrule(
+            self.period_to_rrule_mapping[period_unit],
+            interval=minor_period_count * minor_period_length,
+            dtstart=self._start_dt,
+            until=self._end_dt - relativedelta(seconds=1),
```

Comment:
If the total length of the major period from `start_date` to `end_date` is an exact integer number of minor periods then this is fine. If it's not, though, then the last `TotalActivePeriodsSubscriber` would cover a minor period that extends past `end_date`.  Alternatively, if we did `until=self._end_dt - minor_period_count * minor_period_length` then the last `TotalActivePeriodsSubscriber` would cover the last minor period that fits entirely within the major period (so there would be a gap of less than one minor period at the end, not covered by any `TotalActivePeriodsSubscriber` query).

I'm really not sure which of these two options we would want - the intention is that the major period should consist of an integer number of minor periods, but the parameters allow for that not to be the case. A third option would be to enforce the "integer minor periods per major period" behaviour by getting rid of the `end_date` parameter and replacing it with a `total_minor_periods` parameter (equivalent to the `TotalActivePeriodsSubscriber` parameters for the inner periods) - that would be cleaner, in a way, but from a user perspective it might be nicer to specify the end date than to have to work out the correct number of periods.

Reply:
Having slept on this, I think my suggestion would be to have parameters:
- `start_date`
- `total_major_periods`
- `minor_periods_per_major_period`
- `minor_period_length`
- `period_unit`
(where, in the "active during X distinct hours on Y days over the total period" case, I'm referring to '1 hour' as the minor period and '1 day' as the major period - I'm fully open to other suggestions on names).

When we expose this through the API, we could add a convenience function in flowclient to handle the (probably most commonly used) case of "active X hours on Y days" (i.e. the defaults `period_unit='hour'`, `minor_period_length=1` and `minor_periods_per_major_period=24`). This convenience function could take `start_date` and `end_date` as parameters, and internally calculate the correct value for `total_major_periods`.